### PR TITLE
Add function body to LayerNorm

### PR DESCRIPTION
**Pull request by gramalingam**
**Description**:

(a) Add function body to the schema of LayerNorm
(b) Clarify that the output of LayerNorm is inverse standard-deviation, not inverse variance.

**Motivation and Context**

Function body enables translation of model into ONNX-MLIR.

### CODE REVIEW

**Comment by gramalingam** [comment]
@souptc : this is ready ... please take a look. Thanks!

**Comment by gramalingam** [comment]
@KeDengMS : please take a look, thanks!

File: orttraining/orttraining/test/gradient/function_ops_test.cc [hunk 1]
```diff
@@ -36,244 +21,44 @@ static void RegisterSchemas() {
...
 TEST(SoftmaxGradExpansionTest, NegativeAxis) {
+  RegisterSchemas();
   FunctionTestCase testCase("SoftmaxGrad");
   InitSoftmaxGradTestCase(testCase, {3, 2});
   testCase.AddAttribute("axis", -1);
   testCase.RunTest();
 }
 
 TEST(SoftmaxGradExpansionTest, PositiveAxis) {
+  RegisterSchemas();
```
Comment (KeDengMS) [inline orttraining/orttraining/test/gradient/function_ops_test.cc:46]: >RegisterSchemas [](http://example.com/codeflow?start=2&length=15)

Seems repeating. Could we use fixtures to setup in one class? [https://github.com/google/googletest/blob/master/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests](https://github.com/google/googletest/blob/master/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests)

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -2139,11 +2141,83 @@ Example 4:
...
+            // LayerNormalization <axis, epsilon, stash_type> (X, Scale, B) => (Y, Mean?, InvStdDev?)
+
+            auto* tp = ctx.getInputType(0);
+            if ((tp == nullptr) || (!tp->has_tensor_type()))
+              return false;
+            int64_t T = tp->tensor_type().elem_type();
+
+            auto type_attr = ctx.getAttribute("stash_type");
+            int64_t U = (type_attr != nullptr) ? type_attr->i() : static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+            if ((U != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) && (U != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16))
```
Comment (KeDengMS) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2160]: >((U != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) && (U != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16 [](http://example.com/codeflow?start=15&length=105)

Could you add some comments about why float16 is not supported?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -2139,11 +2141,83 @@ Example 4:
...
+                {{"SuffixShape"}, "ConstantOfShape", {"NumReducedAxes"},                               //
+                 {{"value", mktensor(1)}}},                                                            // [1, ..., 1] for reduced axes
+                {{"ReducedShape"}, "Concat", {"PrefixShape", "SuffixShape"}, {{"axis", int64_t(0)}}},  // [d[0], ..., d[axis-1], 1, ..., 1]
+                {{"X2D"}, "Flatten", {"X"}, {{"axis", axis}}},
+                {{"XU"}, "Cast", {"X2D"}, {{"to", U}}},
+                {{"Mean2D"}, "ReduceMean", {"XU"}, {{"axes", std::vector<int64_t>{1}}}},
+                {{"Square"}, "Mul", {"XU", "XU"}},
+                {{"MeanOfSquare"}, "ReduceMean", {"Square"}, {{"axes", std::vector<int64_t>{1}}}},
+                {{"SquareOfMean"}, "Mul", {"Mean2D", "Mean2D"}},
+                {{"Diff"}, "Sub", {"MeanOfSquare", "SquareOfMean"}},
```
Comment (KeDengMS) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2200]: >Diff [](http://example.com/codeflow?start=19&length=4)

nit: could be named "Variance"

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (gramalingam) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2160]: Good question. I am just following the existing schema, which allows only float and bfloat16, not changing it. Do you think we should change the schema to allow float16 also? But note that this is the type of mean/std-dev, which is allowed to be more precise than the inputs, which may be low-precision. May be we are not using low-precision for mean/std-dev currently?

File: orttraining/orttraining/test/gradient/function_ops_test.cc (hunk 1 above)
Comment (gramalingam) [reply orttraining/orttraining/test/gradient/function_ops_test.cc]: Done

**Review by KeDengMS — APPROVED** [review]

**Pull request merged**